ethstorage/scanner/worker.go

```diff
@@ -0,0 +1,167 @@
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	es "github.com/ethstorage/go-ethstorage/ethstorage"
+)
+
+type Worker struct {
+	sm              *es.StorageManager
+	loadKvFromCache func(uint64, common.Hash) []byte
+	l1              es.Il1Source
+	cfg             Config
+	nextKvIndex     uint64
+	lg              log.Logger
+}
+
+func NewWorker(sm *es.StorageManager, f func(uint64, common.Hash) []byte, l1 es.Il1Source, cfg Config, lg log.Logger) *Worker {
+	return &Worker{
+		sm:              sm,
+		loadKvFromCache: f,
+		l1:              l1,
+		cfg:             cfg,
+		lg:              lg,
+	}
+}
+
+func (s *Worker) ScanBatch(ctx context.Context) error {
+	kvsInBatch, nextKvIndex, err := s.determineBatchIndexRange()
+	if err != nil {
+		return fmt.Errorf("failed to get batch index range: %w", err)
+	}
+	metas, err := s.l1.GetKvMetas(kvsInBatch, rpc.LatestBlockNumber.Int64())
+	if err != nil {
+		s.lg.Error("Scanner: error getting meta range", "kvsInBatch", shortPrt(kvsInBatch))
+		return fmt.Errorf("failed to query KV metas %w", err)
+	}
+	s.lg.Info("Scanner: query KV meta done", "kvsInBatch", shortPrt(kvsInBatch))
+	for i, meta := range metas {
+		select {
+		case <-ctx.Done():
+			s.lg.Info("Scanner canceled, stopping scan", "ctx.Err", ctx.Err())
+			return ctx.Err()
+		default:
+		}
+
+		var err error
+		var found bool
+		var commit common.Hash
+		copy(commit[:], meta[32-es.HashSizeInContract:32])
+		kvIndex := kvsInBatch[i]
+		if s.cfg.Mode == modeCheckMeta {
+			// check meta only
+			var metaLocal []byte
+			metaLocal, found, err = s.sm.TryReadMeta(kvIndex)
+			if metaLocal != nil && !bytes.Equal(metaLocal[0:es.HashSizeInContract], commit[0:es.HashSizeInContract]) {
+				err = es.ErrCommitMismatch
+			}
+		} else if s.cfg.Mode == modeCheckBlob {
+			// query blob and check meta from storage
+			_, found, err = s.sm.TryRead(kvIndex, int(s.sm.MaxKvSize()), commit)
+		} else {
+			panic(fmt.Sprintf("invalid scanner mode: %d", s.cfg.Mode))
+		}
+		if found && err == nil {
+			// happy path
+			s.lg.Debug("Scanner: check KV done", "kvIndex", kvIndex, "commit", commit)
+			continue
+		}
+		if !found {
+			// the shard does not exist locally
+			s.lg.Warn("Scanner: blob not found", "kvIndex", kvIndex, "commit", commit)
+			continue
+		}
+		if err != nil {
+			// chances are the blob is downloaded but not written to the storage
+			if blob := s.loadKvFromCache(kvIndex, commit); blob != nil {
+				s.lg.Debug("Scanner: check KV done from cache", "kvIndex", kvIndex, "commit", commit)
+				continue
+			}
+			s.lg.Error("Scanner: read blob error", "kvIndex", kvIndex, "commit", commit.Hex(), "err", err)
+			if err == es.ErrCommitMismatch {
+				if s.cfg.EsRpc == "" {
+					s.lg.Warn("Scanner: unable to fix blob: no RPC endpoint provided")
+					continue
+				}
+				// try to fix the blob
+				if err := s.fixKv(kvIndex, commit); err != nil {
```

Comment:
Are there any potential race conditions when the downloader performs blob update actions?

Reply:
No. We have lock in StorageManager.TryWrite